File: ios/Sources/KlinikNotificationsFeature/NotificationSettingsScreen.swift
import SwiftUI
import KlinikAPI
import KlinikCore
import KlinikDesign

/// Notification preferences and what was actually delivered (spec M6).
public struct NotificationSettingsScreen: View {
    @Environment(\.colorScheme) private var scheme

    private let model: NotificationSettingsModel

    @State private var state = NotificationSettingsState()

    public init(model: NotificationSettingsModel) {
        self.model = model
    }

    /**
     * The channels a person can turn off.
     *
     * `inApp` is deliberately absent: it is the app's own list, it costs
     * nothing, and a switch that silences the record of what the clinic sent
     * would leave somebody with no way to find out what they missed.
     */
    private static let switchableChannels: [NotificationChannel] = [
        .push, .sms, .email, .whatsapp,
    ]

    public var body: some View {
        VStack(spacing: 0) {
            content
        }
        .background(Tokens.Palette.background.resolve(for: scheme))
        .navigationTitle(L10n.string("notification.settingsTitle"))
        .task { await refresh { await model.load() } }
    }

    @ViewBuilder
    private var content: some View {
        switch state.phase {
        case .loading:
            Spacer()
            ProgressView().accessibilityLabel(L10n.string("common.loading"))
            Spacer()

        case .failed(let message):
            MessageState(
                icon: Tokens.State.labCritical.iconName,
                text: message,
                retryTitle: L10n.string("common.retry")
            ) {
                await refresh { await model.load() }
            }

        case .loaded:
            loaded
        }
    }

    private var loaded: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: Tokens.Spacing.lg) {
                if let error = state.error {
                    ErrorBanner(message: error)
                }

                // Says plainly that turning a channel off does not mean silence:
                // the server falls back to another one for anything clinical.
                Text(L10n.string("notification.fallbackNote"))
                    .font(Tokens.Typography.captionRelative)
                    .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))

                ForEach(NotificationKind.allCases, id: \.rawValue) { kind in
                    PreferenceSection(
                        kind: kind,
                        channels: NotificationSettingsScreen.switchableChannels,
                        state: state
                    ) { channel, enabled in
                        await refresh {
                            await model.set(kind, channel: channel, enabled: enabled)
                        }
                    }
                }

                Divider()

                Text(L10n.string("notification.historyTitle"))
                    .font(Tokens.Typography.subheadingRelative)
                    .foregroundStyle(Tokens.Palette.textPrimary.resolve(for: scheme))

                if state.history.isEmpty {
                    Text(L10n.string("notification.historyEmpty"))
                        .font(Tokens.Typography.bodyRelative)
                        .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
                } else {
                    ForEach(state.history) { delivery in
                        DeliveryRow(delivery: delivery)
                    }
                }
            }
            .padding(Tokens.Spacing.lg)
        }
    }

    private func refresh(_ work: () async -> Void) async {
        await work()
        state = await model.currentState()
    }
}

struct PreferenceSection: View {
    @Environment(\.colorScheme) private var scheme

    let kind: NotificationKind
    let channels: [NotificationChannel]
    let state: NotificationSettingsState
    let onChange: (NotificationChannel, Bool) async -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: Tokens.Spacing.xs) {
            Text(kind.localizedName)
                .font(Tokens.Typography.subheadingRelative)
                .foregroundStyle(Tokens.Palette.textPrimary.resolve(for: scheme))

            ForEach(channels, id: \.rawValue) { channel in
                Toggle(
                    channel.localizedName,
                    isOn: Binding(
                        get: { state.isEnabled(kind, channel) },
                        set: { enabled in Task { await onChange(channel, enabled) } }
                    )
                )
                .disabled(state.saving != nil)
                .font(Tokens.Typography.bodyRelative)
                .frame(minHeight: Tokens.minimumTouchTarget)
            }

            if let quiet = state.quietHours(kind) {
                Text("\(L10n.string("notification.quietHours")): \(quiet.start)–\(quiet.end)")
                    .font(Tokens.Typography.captionRelative)
                    .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
            }
        }
        .padding(.vertical, Tokens.Spacing.xs)
    }
}

/**
 * One delivery attempt.
 *
 * The failure reason and the fallback marker are both shown. A notification
 * that silently failed is how a patient misses a check-up and believes nobody
 * told them — which, on this screen, they can now check.
 */
struct DeliveryRow: View {
    @Environment(\.colorScheme) private var scheme

    let delivery: DeliveredNotification

    var body: some View {
        VStack(alignment: .leading, spacing: Tokens.Spacing.xs) {
            HStack {
                Text(delivery.title)
                    .font(Tokens.Typography.bodyRelative)
                    .foregroundStyle(Tokens.Palette.textPrimary.resolve(for: scheme))

                Spacer()

                Text(delivery.status.localizedName)
                    .font(Tokens.Typography.captionRelative)
                    .foregroundStyle(tint.resolve(for: scheme))
            }

            Text("\(delivery.channel.localizedName) · \(delivery.createdAt.formatted(date: .abbreviated, time: .shortened))")
                .font(Tokens.Typography.captionRelative)
                .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))

            if delivery.isFallback {
                Text(L10n.string("notification.historyFallback"))
                    .font(Tokens.Typography.captionRelative)
                    .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
            }

            if let reason = delivery.failureReason {
                Text(reason)
                    .font(Tokens.Typography.captionRelative)
                    .foregroundStyle(Tokens.Palette.warning.resolve(for: scheme))
            }
        }
        .padding(.vertical, Tokens.Spacing.xs)
        .accessibilityElement(children: .combine)
    }

    private var tint: ThemedColor {
        switch delivery.status {
        case .delivered, .read: return Tokens.Palette.success
        case .failed: return Tokens.Palette.critical
        case .pending, .sent: return Tokens.Palette.info
        }
    }
}

File: android/feature/notifications-ui/src/main/kotlin/xyz/klinik/feature/notifications/ui/NotificationSettingsScreen.kt
package xyz.klinik.feature.notifications.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Surface
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import xyz.klinik.design.Tokens
import xyz.klinik.design.klinikColor
import xyz.klinik.feature.notifications.NotificationSettingsState
import xyz.klinik.feature.notifications.SettingsPhase
import xyz.klinik.network.DeliveredNotification
import xyz.klinik.network.NotificationChannel
import xyz.klinik.network.NotificationDeliveryStatus
import xyz.klinik.network.NotificationKind
import xyz.klinik.network.UiText

data class NotificationStrings(
    val title: String,
    val retry: String,
    val fallbackNote: String,
    val quietHours: String,
    val historyTitle: String,
    val historyEmpty: String,
    val historyFallback: String,
    val kindName: (NotificationKind) -> String,
    val channelName: (NotificationChannel) -> String,
    val statusName: (NotificationDeliveryStatus) -> String,
    val message: (UiText) -> String,
)

/**
 * Notification preferences and what was actually delivered (spec M6).
 *
 * `IN_APP` is deliberately not switchable: it is the app's own list, it costs
 * nothing, and silencing the record of what the clinic sent would leave
 * somebody with no way to find out what they missed.
 */
private val SWITCHABLE = listOf(
    NotificationChannel.PUSH,
    NotificationChannel.SMS,
    NotificationChannel.EMAIL,
    NotificationChannel.WHATSAPP,
)

@Composable
fun NotificationSettingsScreen(
    state: NotificationSettingsState,
    strings: NotificationStrings,
    onRetry: () -> Unit,
    onToggle: (NotificationKind, NotificationChannel, Boolean) -> Unit,
    modifier: Modifier = Modifier,
) {
    Surface(color = klinikColor("background"), modifier = modifier.fillMaxSize()) {
        when (val phase = state.phase) {
            SettingsPhase.Loading -> Centered { CircularProgressIndicator() }

            is SettingsPhase.Failed -> Centered {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Text(strings.message(UiText.Key(phase.messageKey)))
                    TextButton(onClick = onRetry) { Text(strings.retry) }
                }
            }

            SettingsPhase.Loaded -> Column(
                modifier = Modifier
                    .fillMaxSize()
                    .verticalScroll(rememberScrollState())
                    .padding(Tokens.Spacing.lg),
                verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.lg),
            ) {
                state.error?.let { Text(strings.message(it), color = klinikColor("critical")) }

                // Says plainly that turning a channel off is not silence: the
                // server falls back to another one for anything clinical.
                Text(strings.fallbackNote, color = klinikColor("textSecondary"))

                for (kind in NotificationKind.entries) {
                    Column(verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.xs)) {
                        Text(
                            strings.kindName(kind),
                            color = klinikColor("textPrimary"),
                            modifier = Modifier.semantics { heading() },
                        )

                        for (channel in SWITCHABLE) {
                            Row(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .heightIn(min = Tokens.minimumTouchTarget),
                                verticalAlignment = Alignment.CenterVertically,
                            ) {
                                Text(
                                    strings.channelName(channel),
                                    color = klinikColor("textPrimary"),
                                    modifier = Modifier.weight(1f),
                                )

                                Switch(
                                    checked = state.isEnabled(kind, channel),
                                    onCheckedChange = { on -> onToggle(kind, channel, on) },
                                    enabled = state.saving == null,
                                )
                            }
                        }

                        state.quietHours(kind)?.let { (start, end) ->
                            Text(
                                "${strings.quietHours}: $start–$end",
                                color = klinikColor("textSecondary"),
                            )
                        }
                    }
                }

                HorizontalDivider()

                Text(
                    strings.historyTitle,
                    color = klinikColor("textPrimary"),
                    modifier = Modifier.semantics { heading() },
                )

                if (state.history.isEmpty()) {
                    Text(strings.historyEmpty, color = klinikColor("textSecondary"))
                } else {
                    for (delivery in state.history) {
                        DeliveryRow(delivery, strings)
                    }
                }
            }
        }
    }
}

/**
 * One delivery attempt, with the reason it failed if it did.
 *
 * A notification that silently failed is how somebody misses a check-up and
 * believes nobody told them — which, on this screen, they can now check.
 */
@Composable
private fun DeliveryRow(delivery: DeliveredNotification, strings: NotificationStrings) {
    Column(verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.xs)) {
        Row(modifier = Modifier.fillMaxWidth()) {
            Text(
                delivery.title,
                color = klinikColor("textPrimary"),
                modifier = Modifier.weight(1f),
            )

            Text(strings.statusName(delivery.status), color = tintFor(delivery.status))
        }

        Text(
            "${strings.channelName(delivery.channel)} · ${delivery.createdAt}",
            color = klinikColor("textSecondary"),
        )

        if (delivery.isFallback) {
            Text(strings.historyFallback, color = klinikColor("textSecondary"))
        }

        delivery.failureReason?.let {
            Text(it, color = klinikColor("warning"))
        }
    }
}

@Composable
private fun tintFor(status: NotificationDeliveryStatus) = when (status) {
    NotificationDeliveryStatus.DELIVERED, NotificationDeliveryStatus.READ -> klinikColor("success")
    NotificationDeliveryStatus.FAILED -> klinikColor("critical")
    NotificationDeliveryStatus.PENDING, NotificationDeliveryStatus.SENT -> klinikColor("info")
}

@Composable
private fun Centered(content: @Composable () -> Unit) {
    Box(
        modifier = Modifier.fillMaxSize().padding(Tokens.Spacing.xl),
        contentAlignment = Alignment.Center,
    ) { content() }
}
